File: pipeline/tft_pipeline/comp_signature.py
"""Derives a coarse "comp signature" from a match participant, since exact
unit-by-unit compositions are too varied to group literally. A signature is
the participant's top active traits plus their most-invested carry unit —
close enough to how players actually talk about comps ("Bruisers Jinx").
"""

TOP_TRAIT_COUNT = 2


def carry_unit(participant: dict) -> str:
    units = participant.get("units", [])
    if not units:
        return "unknown"

    def item_count(unit: dict) -> int:
        return len(unit.get("itemNames") or unit.get("items") or [])

    return max(units, key=lambda unit: (item_count(unit), unit.get("tier", 0)))[
        "character_id"
    ]


def comp_signature(participant: dict) -> str:
    traits = [t for t in participant.get("traits", []) if t.get("tier_current", 0) > 0]
    traits.sort(key=lambda t: (t["tier_current"], t.get("num_units", 0)), reverse=True)
    top_traits = [t["name"] for t in traits[:TOP_TRAIT_COUNT]]

    return "+".join([*top_traits, carry_unit(participant)])


def item_build(participant: dict, unit_character_id: str) -> str | None:
    for unit in participant.get("units", []):
        if unit.get("character_id") == unit_character_id:
            items = sorted(unit.get("itemNames") or unit.get("items") or [])
            return "+".join(items) if items else None
    return None

File: web/lib/compSignature.ts
/**
 * TypeScript port of pipeline/tft_pipeline/comp_signature.py. Keep the two
 * in sync — this must derive the exact same signature string as the Python
 * pipeline for a given board, or lookups against comp_stats/item_stats
 * won't match.
 */

const TOP_TRAIT_COUNT = 2;

export interface BoardTrait {
  name: string;
  tierCurrent: number;
  numUnits: number;
}

export interface BoardUnit {
  characterId: string;
  itemNames: string[];
  tier: number;
}

export interface BoardState {
  traits: BoardTrait[];
  units: BoardUnit[];
}

export function carryUnit(board: BoardState): string {
  if (board.units.length === 0) {
    return "unknown";
  }

  return board.units.reduce((best, unit) => {
    const bestKey: [number, number] = [best.itemNames.length, best.tier];
    const unitKey: [number, number] = [unit.itemNames.length, unit.tier];
    return unitKey[0] > bestKey[0] || (unitKey[0] === bestKey[0] && unitKey[1] > bestKey[1])
      ? unit
      : best;
  }).characterId;
}

export function compSignature(board: BoardState): string {
  const activeTraits = board.traits
    .filter((t) => t.tierCurrent > 0)
    .sort((a, b) => b.tierCurrent - a.tierCurrent || b.numUnits - a.numUnits)
    .slice(0, TOP_TRAIT_COUNT)
    .map((t) => t.name);

  return [...activeTraits, carryUnit(board)].join("+");
}

export function itemBuild(board: BoardState, unitCharacterId: string): string | null {
  const unit = board.units.find((u) => u.characterId === unitCharacterId);
  if (!unit || unit.itemNames.length === 0) {
    return null;
  }
  return [...unit.itemNames].sort().join("+");
}
